src/auth/src/credentials.rs

```diff
@@ -253,13 +241,56 @@ pub async fn create_access_token_credential() -> Result<Credential> {
     }
 }
 
+#[derive(Debug, PartialEq)]
+enum AdcPath {
+    FromEnv(String),
+    WellKnown(String),
+    None,
+}
+
+#[derive(Debug, PartialEq)]
+enum AdcContents {
+    Contents(String),
+    FallbackToMds,
+}
+
+fn load_adc() -> Result<AdcContents> {
+    match adc_path() {
+        AdcPath::None => Ok(AdcContents::FallbackToMds),
+        AdcPath::FromEnv(path) => {
+            match std::fs::read_to_string(&path) {
+                Ok(contents) => Ok(AdcContents::Contents(contents)),
+                Err(e) if e.kind() == std::io::ErrorKind::NotFound => Err(CredentialError::new(
+                    false,
+                    Box::from(format!(
+                        "Failed to load Application Default Credentials (ADC) from {path}. Check that the `GOOGLE_APPLICATION_CREDENTIALS` environment variable points to a valid file."
+                    )))),
+                Err(e) => Err(CredentialError::new(false, e.into())),
+            }
+        },
+        AdcPath::WellKnown(path) => {
+            match std::fs::read_to_string(path) {
+                Ok(contents) => Ok(AdcContents::Contents(contents)),
+                Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(AdcContents::FallbackToMds),
+                Err(e) => Err(CredentialError::new(false, e.into())),
+            }
+        },
+    }
+}
+
 /// The path to Application Default Credentials (ADC), as specified in [AIP-4110].
 ///
 /// [AIP-4110]: https://google.aip.dev/auth/4110
-fn adc_path() -> Option<String> {
+fn adc_path() -> AdcPath {
     std::env::var("GOOGLE_APPLICATION_CREDENTIALS")
         .ok()
-        .or_else(adc_well_known_path)
+        .map_or_else(
+            || match adc_well_known_path() {
```

Comment:
🫤  maybe a bit too big for a `.map_or_else()`

Reply:
Rearranged.